File: cpp/23_file_handling.cpp
// ============================================================
// Topic:  File Handling  (FOCUS: fstream, ofstream, ifstream, append, remove)
// Summary: Write, read, append, existence check, line count, delete — raw fstream.
// Run:     g++ -std=c++17 cpp/23_file_handling.cpp -o cpp/23_file_handling && cpp/23_file_handling
//
// KEY DIFFERENCES:
//   C++:     fstream / ofstream / ifstream, std::ios::app, std::remove (cstdio)
//   Python:  open(), read(), write(), os.remove()
//   Java:    Files, BufferedReader/Writer, Files.deleteIfExists()
//   C#:      File, StreamReader/Writer, File.Delete()
//   Rust:    std::fs::File, OpenOptions::append(), std::fs::remove_file()
//   TS/JS:   fs.readFileSync / writeFileSync / appendFileSync / unlinkSync
// ============================================================

#include <cstdio>
#include <fstream>
#include <iostream>
#include <string>

int main() {
    const std::string path = "_alllangs_demo.txt";

    // ----- 1. Write with ofstream -----
    {
        std::ofstream out(path);
        out << "hello file";
        std::cout << "write: " << (out.good() ? "ok" : "fail") << std::endl;
    }

    // ----- 2. Read with ifstream -----
    {
        std::ifstream in(path);
        std::string content;
        std::getline(in, content);
        std::cout << "read: " << content << std::endl;
    }

    // ----- 3. Append with fstream in append mode -----
    {
        std::fstream file(path, std::ios::out | std::ios::app);
        file << "!";
        std::cout << "append: " << (file.good() ? "ok" : "fail") << std::endl;
    }

    // ----- 4. Read again with ifstream -----
    {
        std::ifstream in(path);
        std::string content;
        std::getline(in, content);
        std::cout << "read: " << content << std::endl;
    }

    // ----- 5. Existence check with ifstream -----
    {
        std::ifstream check(path);
        std::cout << "exists: " << (check.good() ? "true" : "false") << std::endl;
    }

    // ----- 6. Line count with ifstream -----
    {
        std::ifstream in(path);
        int lines = 0;
        std::string line;
        while (std::getline(in, line)) {
            ++lines;
        }
        std::cout << "lines: " << lines << std::endl;
    }

    // ----- 7. Delete with remove() -----
    {
        const int result = std::remove(path.c_str());
        std::cout << "deleted: " << (result == 0 ? "ok" : "fail") << std::endl;
    }

    return 0;
}

File: rust/23_file_handling.rs
// ============================================================
// Topic:  File Handling
// Summary: fs::write, read_to_string, OpenOptions append, metadata, remove_file.
// Run:     rustc --edition 2021 rust/23_file_handling.rs -o rust/23_file_handling.exe && rust/23_file_handling.exe
//
// KEY DIFFERENCES:
//   Python:  open(), read(), write(), pathlib.Path
//   TS/JS:   fs.readFileSync, fs.writeFileSync, fs.appendFileSync
//   Rust:    std::fs + std::io; Result for I/O errors; Path/PathBuf
//   Java:    Files.readString, Files.write, BufferedWriter
//   C++:     fstream, std::filesystem
// ============================================================

use std::fs::{self, OpenOptions};
use std::io::Write;
use std::path::PathBuf;

fn main() {
    let path: PathBuf = std::env::temp_dir().join("_alllangs_demo.txt");

    // Remove leftover from a previous run (ignore if missing)
    let _ = fs::remove_file(&path);

    // ----- 1. Write entire file (creates or overwrites) -----
    fs::write(&path, "hello file").expect("write failed");
    println!("write: ok");

    // ----- 2. Read whole file as String -----
    let content = fs::read_to_string(&path).expect("read failed");
    println!("read: {}", content.trim());

    // ----- 3. Append with OpenOptions -----
    let mut file = OpenOptions::new()
        .append(true)
        .open(&path)
        .expect("open for append failed");
    file.write_all(b"!").expect("append failed");
    println!("append: ok");

    // ----- 4. Read again after append -----
    let content = fs::read_to_string(&path).expect("read failed");
    println!("read: {}", content.trim());

    // ----- 5. metadata — check file exists -----
    let exists = fs::metadata(&path).is_ok();
    println!("exists: {}", exists);

    // ----- 6. Line count from content -----
    let lines = content.lines().count();
    println!("lines: {}", lines);

    // ----- 7. Delete file -----
    fs::remove_file(&path).expect("remove failed");
    println!("deleted: ok");
}